symengine/utilities/matchpycpp/substitution.h

```diff
@@ -0,0 +1,83 @@
+#ifndef SYMENGINE_UTILITIES_MATCHPYCPP_SUBSTITUTION_H_
+#define SYMENGINE_UTILITIES_MATCHPYCPP_SUBSTITUTION_H_
+
+#include <symengine/basic.h>
+#include <symengine/add.h>
+#include <symengine/pow.h>
+#include <symengine/mul.h>
+
+#include <map>
+#include <vector>
+
+using namespace std;
+using namespace SymEngine;
+
+
+struct lessBasic {
+    bool operator()(const RCP<const Basic> &x, const RCP<const Basic> &y) const
+    {
+        return (x->__hash__()) < (y->__hash__());
+    }
+};
+
+bool operator==(RCP<const Basic> &a, RCP<const Basic> &b)
+{
+    return (a->__hash__()) == (b->__hash__());
+}
+
+typedef multiset<RCP<const Basic>, lessBasic> MultisetOfBasic;
+
+typedef map<string, MultisetOfBasic> Substitution;
+
+int try_add_variable(Substitution &subst, const string variable_name,
+                     const MultisetOfBasic &replacement)
+{
+    if (subst.find(variable_name) == subst.end()) {
+        subst[variable_name] = replacement;
```

Comment:
It's better to store the value of `subst.find(variable_name)` and it use it as a hint for insertion. Avoids a second tree traversal